src/drivers/wayland/wayland.c

```diff
@@ -0,0 +1,2555 @@
+        obj->height, obj->window->frame_counter);
+    (void)obj;
+    smm_release_buffer((smm_buffer_t*)data);
+}
+
+static const struct wl_buffer_listener wl_buffer_listener = {
+    .release = handle_buffer_release_wl_cb,
+};
+
+static void cache_clear(lw_window_t* window)
+{
+    window->dmg_cache.start = window->dmg_cache.end;
+    window->dmg_cache.size = 0;
+}
+
+static void cache_purge(lw_window_t* window, smm_buffer_t* buf)
+{
+    lv_area_t* next_dmg;
+    smm_buffer_t* next_buf = smm_next_buffer(buf);
+
+    /* Remove all damage areas up until start of next buffers damage */
+    if (next_buf == NULL) {
+        cache_clear(window);
+    }
+    else {
+        next_dmg = SMM_BUFFER_PROPERTIES(next_buf)->tag[LW_TAG_BUFFER_DAMAGE];
+        while ((window->dmg_cache.cache + window->dmg_cache.start) != next_dmg) {
+            window->dmg_cache.start++;
+            window->dmg_cache.start %= LW_DMG_CACHE_CAPACITY;
+            window->dmg_cache.size--;
+        }
+    }
+}
+
+static void cache_add_area(lw_window_t* window, smm_buffer_t* buf, const lv_area_t* area)
+{
+    if (SMM_BUFFER_PROPERTIES(buf)->tag[LW_TAG_BUFFER_DAMAGE] == NULL) {
+        /* Buffer damage beyond cache capacity */
+        return;
+    }
+
+    if ((window->dmg_cache.start == window->dmg_cache.end) &&
+        (window->dmg_cache.size)) {
+        /* This buffer has more damage then the cache's capacity, so
+         * clear cache and leave buffer damage unrecorded
+         */
+        cache_clear(window);
+        SMM_TAG(buf, LW_TAG_BUFFER_DAMAGE, NULL);
+        return;
+    }
+
+    /* Add damage area to cache */
+    lv_memcpy(window->dmg_cache.cache + window->dmg_cache.end,
+        area,
+        sizeof(lv_area_t));
+
+    window->dmg_cache.end++;
+    window->dmg_cache.end %= LW_DMG_CACHE_CAPACITY;
+    window->dmg_cache.size++;
+}
+
+static void cache_apply_areas(lw_window_t* window, void* dest, void* src, smm_buffer_t* src_buf)
+{
+    unsigned long offset;
+    unsigned char start;
+    int y;
+    lv_area_t* dmg;
+    lv_area_t* next_dmg;
+    smm_buffer_t* next_buf = smm_next_buffer(src_buf);
+    const struct smm_buffer_properties* props = SMM_BUFFER_PROPERTIES(src_buf);
+    struct lw_surface_object* obj = SMM_GROUP_PROPERTIES(props->group)->tag[LW_TAG_LOCAL];
+
+    if (next_buf == NULL) {
+        next_dmg = (window->dmg_cache.cache + window->dmg_cache.end);
+    }
+    else {
+        next_dmg = SMM_BUFFER_PROPERTIES(next_buf)->tag[LW_TAG_BUFFER_DAMAGE];
+    }
+
+    /* Apply all buffer damage areas */
+    start = ((lv_area_t*)SMM_BUFFER_PROPERTIES(src_buf)->tag[LW_TAG_BUFFER_DAMAGE] - window->dmg_cache.cache);
+    while ((window->dmg_cache.cache + start) != next_dmg) {
+        /* Copy an area from source to destination (line-by-line) */
+        dmg = (window->dmg_cache.cache + start);
+        for (y = dmg->y1; y <= dmg->y2; y++) {
+            offset = (dmg->x1 + (y * (obj->width * LW_COLOR_PX_SIZE)));
```

Comment:
There is a pointer arithmetic issue here, that leads to glitches on weston, when running the widgets demo the menu buttons are not rendered correctly.

Reply:
with mutter (wayland 1.23, protocol 1.36),  no problem.